### PR TITLE
planner: add binding and recursive test for `Merge` hint

### LINKED ISSUE #17472: support common table expression (CTE)
## Description

**Is your feature request related to a problem? Please describe:**

Related issues:
- https://github.com/pingcap/tidb/issues/6824 request to support recursive common table expression
- https://github.com/pingcap/tidb/issues/17916 request to support common table expression to run TPC-DS and other analytic workloads
- https://github.com/pingcap/parser/issues/289 request to support common table expression to be compatible with MySQL 8.0

**Describe the feature you'd like:**

Support the common table expression feature like MySQL [13.2.15 WITH (Common Table Expressions)](https://dev.mysql.com/doc/refman/8.0/en/with.html):

> A common table expression (CTE) is a named temporary result set that exists within the scope of a single statement and that can be referred to later within that statement, possibly multiple times. The following discussion describes how to write statements that use CTEs.

There are two kinds of CTEs:

- non-recursive CTE
- recursive CTE

A typical query example of non-recursive CTE:

```sql
WITH
  cte1 AS (SELECT a, b FROM table1),
  cte2 AS (SELECT c, d FROM table2)
SELECT b, d FROM cte1 JOIN cte2
WHERE cte1.a = cte2.c;
```

A typical query example of recursive CTE:

```sql
WITH RECURSIVE cte (n) AS
(
  SELECT 1
  UNION ALL
  SELECT n + 1 FROM cte WHERE n < 5
)
SELECT * FROM cte;
```

### PR DESCRIPTION
**Pull request by dayicklp**
### What problem does this PR solve?

Issue Number: close #17472

Problem Summary: add binding and recursive test for `Merge` hint(#34574 ), and fix some unit test and function.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -4010,11 +4016,6 @@ func (b *PlanBuilder) buildSelect(ctx context.Context, sel *ast.SelectStmt) (p L
 	}
 
 	// Determines whether to use the Merge hint in a CTE query.
```
Comment (winoros) [inline planner/core/logical_plan_builder.go:4018]: The comment is not moved

File: planner/core/logical_plan_builder.go [hunk 2]
```diff
@@ -3631,6 +3631,11 @@ func (b *PlanBuilder) pushTableHints(hints []*ast.TableOptimizerHint, currentLev
 			limitHints.preferLimitToCop = true
 		case HintMerge:
 			MergeHints.preferMerge = true
+			if hint.Tables != nil {
+				b.ctx.GetSessionVars().StmtCtx.AppendWarning(ErrInternal.GenWithStack("The MERGE hint is not used correctly, maybe it inputs a table name."))
+				MergeHints.preferMerge = false
```
Comment (wjhuang2016) [inline planner/core/logical_plan_builder.go:3636]: You can put `MergeHints.preferMerge = true` after L3638, and remove this line.

Comment (dayicklp) [reply planner/core/logical_plan_builder.go:3636]: ok

**Review by winoros — APPROVED** [review]

**Pull request merged**